src/main/java/net/snowflake/ingest/utils/ParameterProvider.java

```diff
@@ -149,20 +160,74 @@ private void setParameterMap(Map<String, Object> parameterOverrides, Properties
 
     this.updateValue(
         MAX_CHUNK_SIZE_IN_BYTES, MAX_CHUNK_SIZE_IN_BYTES_DEFAULT, parameterOverrides, props);
+
+    this.updateValue(MAX_CLIENT_LAG, MAX_CLIENT_LAG_DEFAULT, parameterOverrides, props);
+    this.updateValue(
+        MAX_CLIENT_LAG_ENABLED, MAX_CLIENT_LAG_ENABLED_DEFAULT, parameterOverrides, props);
   }
 
-  /** @return Longest interval in milliseconds between buffer flushes */
+  /**
+   * @return Longest interval in milliseconds between buffer flushes
+   */
   public long getBufferFlushIntervalInMs() {
-    Object val =
-        this.parameterMap.getOrDefault(
-            BUFFER_FLUSH_INTERVAL_IN_MILLIS, BUFFER_FLUSH_INTERVAL_IN_MILLIS_DEFAULT);
-    if (val instanceof String) {
-      return Long.parseLong(val.toString());
+    if (getMaxClientLagEnabled()) {
+      if (cachedBufferFlushIntervalMs != -1L) {
+        return cachedBufferFlushIntervalMs;
+      }
+      long lag = getMaxClientLagMs();
+      if (cachedBufferFlushIntervalMs == -1L) {
+        cachedBufferFlushIntervalMs = lag;
+      }
+      return cachedBufferFlushIntervalMs;
+    } else {
+      Object val =
+          this.parameterMap.getOrDefault(
+              BUFFER_FLUSH_INTERVAL_IN_MILLIS, BUFFER_FLUSH_INTERVAL_IN_MILLIS_DEFAULT);
+      return (val instanceof String) ? Long.parseLong(val.toString()) : (long) val;
+    }
+  }
+
+  private long getMaxClientLagMs() {
```

Comment:
This method might cause some confusing results if the customer enters wrong inputs.

(happy case) if the input is "1 second" the output is 1000

(not string) if the input is 1 or 5 or 1000 the result is 1000
(no unit) if the input is "1" or "5" or "1000" the result is 1000
(double space) if the input is "5 seconds" or "abc abc" the result is 1000

(non parsable value) if the input is "abc abc" then we get an IllegalArgumentException

Reply:
Should we also throw IllegalArgumentException for the above cases instead of reverting back to default?

Reply:
Yeah let's throw an exception here.